feat(identity): how to use auth servers, principals, and OIDC (#5702)
Kong/developer.konghq.com: app/_how-tos/gateway/authenticate-oauth-clients-with-kong-identity.md

---
title: Authenticate OAuth clients with a Kong Identity authorization server
permalink: /how-to/authenticate-oauth-clients-with-kong-identity/
content_type: how_to
breadcrumbs:
  - /identity/
related_resources:

>   - text: Authentication
>     url: /gateway/authentication/

Reviewer: ```suggestion
  - text: Authentication
    url: /gateway/authentication/
  - text: "{{site.identity}}"
    url: /identity/
  - text: "{{site.identity}} principals and directories"
    url: /identity/principals
```

description: Use the OpenID Connect plugin to map authenticated OAuth clients to Kong Identity principals.
products:
    - gateway
    - identity

plugins:
  - openid-connect
works_on:
    - konnect

>     - on-prem

Reviewer: ```suggestion
```
Since we're doing principal lookup, only Konnect should be supported.

Author: The principal lookup doesn't gate this flow, so I've clarified the docs to reflect that. T[he client creds + auth server flow is actually compatible with on prem](https://kongstrong.slack.com/archives/C0A2G9U5CTX/p1780500478661149?thread_ts=1780499980.991149&cid=C0A2G9U5CTX).

      icon_url: /assets/icons/identity.svg

cleanup:
  inline:
    - title: Clean up Konnect environment
      include_content: cleanup/platform/konnect
      icon_url: /assets/icons/gateway.svg
    - title: Destroy the {{site.base_gateway}} container
      include_content: cleanup/products/gateway
      icon_url: /assets/icons/gateway.svg
---

## Create an authorization server in {{site.identity}}

An authorization server in {{site.identity}} issues the OAuth tokens that clients present to authenticate to your service.

Create an authorization server using the [`/v1/auth-servers` endpoint](/api/konnect/kong-identity/v1/#/operations/createAuthServer):

<!--vale off-->
{% konnect_api_request %}
url: /v1/auth-servers
status_code: 201
method: POST
headers:
  - 'Content-Type: application/json'
body:
  name: "example-auth-server"
  description: "Example auth server"
  audience: "orders-api"
extract_body:
  - name: 'id'
    variable: AUTH_SERVER_ID
  - name: 'issuer'
    variable: ISSUER_URL
capture:
  - variable: AUTH_SERVER_ID
    jq: ".id"
  - variable: ISSUER_URL
    jq: ".issuer"
{% endkonnect_api_request %}
<!--vale on-->

## Create a client

> The client is the credential that the OpenID Connect plugin uses to authenticate. {{site.konnect_short_name}} autogenerates the client ID and secret.

Reviewer: ```suggestion
The client is the credential that the OpenID Connect plugin uses to authenticate. {{site.identity}} autogenerates the client ID and secret.
```

Add the identity using the `/v2/directories/$DIRECTORY_ID/principals/$PRINCIPAL_ID/identities` endpoint:

<!--vale off-->
{% konnect_api_request %}
url: /v2/directories/$DIRECTORY_ID/principals/$PRINCIPAL_ID/identities
status_code: 201
method: POST
headers:
  - 'Content-Type: application/json'
  - 'Authorization: Bearer $KONNECT_TOKEN'
body:
  type: auth_server_client
  auth_server_id: $AUTH_SERVER_ID
  client_id: $CLIENT_ID
{% endkonnect_api_request %}
<!--vale on-->

## Get the directory name

To configure the OpenID Connect plugin, you need the name of the directory you created. Store it as `DECK_DIRECTORY_NAME` so decK can read it during sync:

{% include /how-tos/steps/get-directory-name.md %}

## Generate a salt token

{% include how-tos/steps/deck-salt-token.md %}

## Configure the OpenID Connect plugin

Export the variables to decK:

```bash
export DECK_ISSUER_URL=$ISSUER_URL
export DECK_CLIENT_ID=$CLIENT_ID
export DECK_CLIENT_SECRET=$CLIENT_SECRET
```

> Configure the [OpenID Connect plugin](/plugins/openid-connect/) to use {{site.identity}} as the identity provider and look up principals in your directory. Setting `principals.enabled` to `true` maps the authenticated token to a principal in the directory. This example applies the plugin to the `example-service` you created in the prerequisites.

Author: ```suggestion
Configure the [OpenID Connect plugin](/plugins/openid-connect/) to use {{site.identity}} as the identity provider and map the client credentials. Setting `principals.enabled` to `true` maps the authenticated token to a directory. This example applies the plugin to the `example-service` you created in the prerequisites.
```

---
title: Authenticate OAuth clients with a Kong Identity authorization server
permalink: /how-to/authenticate-oauth-clients-with-kong-identity/
content_type: how_to
breadcrumbs:
  - /identity/
related_resources:
  - text: Authentication
    url: /gateway/authentication/
  - text: "{{site.identity}}"
    url: /identity/
  - text: "{{site.identity}} principals and directories"

>     url: /identity/principals

Author: ```suggestion
    url: /identity/principals/
```

  - text: Authentication
    url: /gateway/authentication/

description: Use the OpenID Connect plugin to map authenticated OAuth clients to Kong Identity principals.
products:
    - gateway
    - identity

plugins:
  - openid-connect
works_on:

>     - konnect

Author: ```suggestion
    - konnect
    - on-prem
    
```
